lib/msf/core/windows_version.rb

```diff
@@ -0,0 +1,160 @@
+    Server2003_SP2 = Rex::Version.new('5.2.3790.2')
+    Vista_SP0 = Server2008_SP0 = Rex::Version.new('6.0.6000.0')
+    Vista_SP1 = Server2008_SP1 = Rex::Version.new('6.0.6001.1')
+    Vista_SP2 = Server2008_SP2 = Rex::Version.new('6.0.6002.2')
+    Win7_SP0 = Server2008_R2_SP0 = Rex::Version.new('6.1.7600.0')
+    Win7_SP1 = Server2008_R2_SP1 = Rex::Version.new('6.1.7601.1')
+    Win8 = Server2012 = Rex::Version.new('6.2.9200.0')
+    Win81 = Server2012_R2 = Rex::Version.new('6.3.9600.0')
+    Win10_1507 = Win10_InitialRelease = Rex::Version.new('10.0.10240.0')
+    Win10_1511 = Rex::Version.new('10.0.10586.0')
+    Win10_1607 = Server2016 = Rex::Version.new('10.0.14393.0')
+    Win10_1703 = Rex::Version.new('10.0.15063.0')
+    Win10_1709 = Rex::Version.new('10.0.16299.0')
+    Win10_1803 = Rex::Version.new('10.0.17134.0')
+    Win10_1809 = Server2019 = Rex::Version.new('10.0.17763.0')
+    Win10_1903 = Rex::Version.new('10.0.18362.0')
+    Win10_1909 = Rex::Version.new('10.0.18363.0')
+    Win10_2004 = Rex::Version.new('10.0.19041.0')
+    Win10_20H2 = Rex::Version.new('10.0.19042.0')
+    Win10_21H1 = Rex::Version.new('10.0.19043.0')
+    Win10_21H2 = Rex::Version.new('10.0.19044.0')
+    Win10_22H2 = Rex::Version.new('10.0.19044.0')
+    Server2022 = Rex::Version.new('10.0.20348.0')
+    Win11_21H2 = Rex::Version.new('10.0.22000.0')
+    Win11_22H2 = Rex::Version.new('10.0.22621.0')
+
+    module MajorRelease
+      NT351 = "Windows NT 3.51"
+      Win95 = "Windows 95"
+      Win98 = "Windows 98"
+      WinME = "Windows ME"
+
+      XP = "Windows XP"
+      Server2003 = "Windows Server 2003"
+
+      Vista = "Windows Vista"
+      Server2008 = "Windows Server 2008"
+      
+      Win7 = "Windows 7"
+      Server2008R2 = "Windows 2008 R2"
+
+      Win8 = "Windows 8"
+      Server2012 = "Windows Server 2012"
+
+      Win81 = "Windows 8.1"
+      Server2012R2 = "Windows Server 2012 R2"
+
+      Win10Plus = "Windows 10+"
+      Server2016Plus = "Windows Server 2016+"
+    end
+
+    def initialize(major, minor, build, service_pack, product_type)
+      self._major = major
+      self._minor = minor
+      self._build = build
+      self._service_pack = service_pack
+      self.product_type = product_type
+    end
+
+    # The specific build number of this version (major.minor.build.service_pack)
+    def build_number
+      Rex::Version.new("#{_major}.#{_minor}.#{_build}.#{_service_pack}")
+    end
+
+    # Is this OS a Windows Server instance?
+    def is_windows_server
+      # There are other types than just workstation/server/DC, but Microsoft's own documentation says
+      # "If it's not Workstation, then it's Server"
+      # https://learn.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-osversioninfoexa
+      self.product_type != VER_NT_WORKSTATION
+    end
+
+    # This Windows Server has been promoted to a DC
+    def is_domain_controller
+      self.product_type == VER_NT_DOMAIN_CONTROLLER
+    end
```

Comment:
Convention for Boolean methods is to use `?`. For example:

```suggestion
    # This Windows Server has been promoted to a DC
    def is_domain_controller?
      self.product_type == VER_NT_DOMAIN_CONTROLLER
    end
```

This also applies to a few other methods in this library.

Whether to preface the method name with `is_` is also questionable but debated. ie: `def domain_controller?`